packages/material-ui-utils/src/elementTypeAcceptingRef.js

```diff
@@ -0,0 +1,50 @@
+import * as PropTypes from 'prop-types';
+import React from 'react';
+import { isLazy, isMemo } from 'react-is';
+import chainPropTypes from './chainPropTypes';
+
+function isClassComponent(elementType) {
+  // elementType.prototype?.isReactComponent
+  const { prototype = {} } = elementType;
+
+  return Boolean(prototype.isReactComponent);
+}
+
+function elementTypeAcceptingRef(props, propName, componentName, location, propFullName) {
+  const propValue = props[propName];
+  const safePropName = propFullName || propName;
+
+  if (propValue == null) {
+    return null;
+  }
+
+  let warningHint;
+
+  /**
+   * blacklisting instead of whitelisting
+   *
+   * blacklisting will miss some components like React.Fragment. Those will at least
+   * trigger a warning in react.
+   * We can't whitelist because there is no safe way to detect React.forwardRef
+   * or class components. Safe means there's no public API.
+   *
+   */
+  if (isLazy(propValue)) {
+    warningHint = 'But you passed a React.lazy component.';
+  } else if (isMemo(propValue)) {
+    warningHint = 'But you passed a React.memo component.';
+  } else if (typeof propValue === 'function' && !isClassComponent(propValue)) {
+    warningHint = 'Did you accidentally provide a plain function component instead?';
+  }
+
+  if (warningHint !== undefined) {
+    return new Error(
+      `Invalid ${location} \`${safePropName}\` supplied to \`${componentName}\`. ` +
+        `Expected an element type that can hold a ref. ${warningHint}`,
+    );
+  }
+
+  return null;
+}
+
+export default chainPropTypes(PropTypes.elementType, elementTypeAcceptingRef);
```

Comment:
I'm not sure about this prop type strategy efficiency. What do you think of?
- React should raise a warning when people are providing a function component. Users will create issues this. Hopefully, with GitHub's issue deduplication & search, we can have a single reference issue and point to a documentation section explaining what to do.
- We will need to remove findDOMNode() at some point. We can create a new warning at the reference consumption, and we can point people to a dedicated documentation section when we receive a non-host element reference.

Reply:
You think it's better to let users find out if that prop type warning is an isse on our side or theirs? Then have to go to the github repo, search for the issue, hope they put in the right keywords, read the issue until they find the section that would be immediately linked from the prop type warning proposed in this PR?

Reply:
> You think it's better to let users find out if that prop type warning is an issue on our side or theirs?

It's not better. I'm trying to estimate the fraction of the people that will do that, in the hypothesis of releasing the proptype. It's it's 80%, what's the point? If it's only 20%, then its awesome.

Reply:
So you're concern is about making this public? Could mark this as private in the code. We don't have a documentation for `@material-ui/utils` anyway.

Reply:
My concern was with the efficiency of the proptype. But I think that it can help, even a little can be great here given how time-consuming a problem can be for the users. Man, imagine we had to bubble up the DOM events manually 🙈. It would be so error-prone, it's what we are doing with the ref.